rmf_site_editor/src/interaction/camera_controls/cursor.rs

```diff
@@ -0,0 +1,385 @@
+            let orbit_sensitivity = 1.0;
+            let delta_x =
+                cursor_motion.x / window_size.x * std::f32::consts::PI * orbit_sensitivity;
+            let delta_y =
+                cursor_motion.y / window_size.y * std::f32::consts::PI * orbit_sensitivity;
+            let yaw = Quat::from_rotation_z(-delta_x);
+            let pitch = Quat::from_rotation_x(-delta_y);
+
+            // Get target rotation, excluding pitch if upside down
+            let mut target_transform = camera_transform.clone();
+            target_transform.rotation = (yaw * camera_transform.rotation) * pitch;
+            if target_transform.up().dot(Vec3::Z) <= 0.0 {
+                target_transform.rotation = yaw * camera_transform.rotation;
+            };
+
+            // Calculate translation to orbit around camera centre
+            let orbit_radius = orbit_radius - 0.5 * scroll_motion;
+            let target_rotation = Mat3::from_quat(target_transform.rotation);
+            target_transform.translation =
+                orbit_center + target_rotation.mul_vec3(Vec3::new(0.0, 0.0, orbit_radius));
+
+            // Get the rotation difference to be multiplied into the current rotation
+            let start_rotation = Mat3::from_quat(camera_transform.rotation);
+            cursor_command.rotation_delta =
+                Quat::from_mat3(&(start_rotation.inverse() * target_rotation));
+            cursor_command.translation_delta =
+                target_transform.translation - camera_transform.translation;
+            is_cursor_selecting = true;
+        }
+        CameraCommandType::Inactive => (),
+    }
+
+    cursor_command.command_type = command_type;
+    cursor_command.cursor_selection = if is_cursor_selecting {
+        Some(cursor_selection)
+    } else {
+        None
+    };
+
+    return cursor_command;
+}
+
+// Returns the object selected by the cursor, if none, defaults to ground plane or arbitrary point in front
+fn get_cursor_selected_point(cursor_raycast_source: &RaycastSource<SiteRaycastSet>) -> Vec3 {
+    let cursor_ray = cursor_raycast_source.get_ray().unwrap();
+
+    match cursor_raycast_source.get_nearest_intersection() {
+        Some((_, intersection)) => intersection.position(),
+        None => {
+            // If valid intersection with groundplane
+            let pitch = cursor_ray.direction().z.acos().to_degrees() - 90.0;
+            let denom = Vec3::Z.dot(cursor_ray.direction());
+            if denom.abs() > f32::EPSILON && pitch.abs() >= MIN_SELECTION_ANGLE {
+                let dist = (-1.0 * cursor_ray.origin()).dot(Vec3::Z) / denom;
+                if dist > f32::EPSILON {
+                    return cursor_ray.origin() + cursor_ray.direction() * dist;
+                }
+            }
+
+            // No groundplane intersection
+            // Pick a point of a virtual sphere around the camera, of same radius as its height
```

Comment:
Using the height of the camera as the radius of the maximum sphere is an interesting choice. Is there a specific rationale behind that, or is it just based on experimenting with different ideas and seeing what feels best?

Reply:
My intuition was that when the cursor is grabbing nothing, it should behave as if selecting the nearest object, which would be the ground. In hindsight, I think a better way to approximate the nearest object would be to grab the nearest [visible entity](https://docs.rs/bevy/latest/bevy/core_pipeline/core_3d/struct.Camera3dBundle.html#structfield.visible_entities) from the active camera